src/bootloader/bootloader.cpp

```diff
@@ -0,0 +1,286 @@
+            return false;
+        }
+
+        log(LOG_SEVERITY_DEBUG, "Programming 0x%08X (size %zu)", address, read);
+        if (!flash_program(address, buffer, read)) {
+            log(LOG_SEVERITY_ERROR, "Writing the bootloader to FLASH failed", errno);
+            return false;
+        }
+        log(LOG_SEVERITY_DEBUG, "Programming 0x%08X (size %zu) finished", address, read);
+
+        address += read;
+        copied_bytes += read;
+
+        progress(copied_bytes);
+    }
+
+    return true;
+}
+
+/// Copy the bootloader from file to FLASH
+///
+/// The bootloader is located in the first 128 KB of FLASH
+/// and consists of the preboot (the first 16 KB) and the main bootloader (the rest).
+///
+/// The FLASH is organised into sectors and before we write anything, we must erase the whole sector
+/// sector0 - 16 KB : stores the preboot
+/// sector1 - 16 KB : main bootloader
+/// sector2 - 16 KB : main bootloader
+/// sector3 - 16 KB : main bootloader
+/// sector4 - 64 KB : main bootloader
+///
+template <typename ProgressCallback>
+static void copy_bootloader_to_flash(FILE *bootloader_bin, ProgressCallback progress) {
+    const size_t total_bytes = 131072;
+    size_t bytes_in_preceding_sectors = 0;
+
+    for (unsigned sector = 0; sector < bootloader_sector_count; sector++) {
+
+        // do not reflash preboot if not necessary
+        if (sector == 0) {
+            uint32_t expected_preboot_crc = 0;
+            if (!calculate_file_crc(bootloader_bin, bootloader_sector_get_size(0), expected_preboot_crc)) {
+                fatal_error("expected preboot crc calculation failed");
+            }
+
+            uint32_t current_preboot_crc = crc32_calc(bootloader_sector_get_address(0), bootloader_sector_get_size(0));
+            if (current_preboot_crc == expected_preboot_crc && false) {
+                log(LOG_SEVERITY_INFO, "No need to update preboot. Skipping sector 0.");
+                continue;
+            } else {
+                log(LOG_SEVERITY_INFO, "Going to update preboot now.");
+            }
+        }
+
+        log(LOG_SEVERITY_INFO, "Flashing sector %i", sector);
+
+        // add random delay to make preboot flashing less predictable
+        if (sector == 0) {
+            uint32_t delay_ms = 100 + (random_number() % 7000);
+            osDelay(delay_ms);
+        }
+
+        // seek at the sector in the file
+        if (fseek(bootloader_bin, bootloader_sector_get_address(sector) - bootloader_sector_get_address(0), SEEK_SET) != 0) {
+            fatal_error("bootloader update: failed to seek sector");
+        }
+
+        // erase the sector
+        HAL_FLASH_Unlock();
+        if (!flash_erase_sector(sector)) {
+            fatal_error("bootloader update: failed to erase sector");
+        }
+
+        // program the sector
+        HAL_FLASH_Unlock();
+        bool flash_successful = flash_program_sector(sector, bootloader_bin, [&](size_t bytes_written) {
+            if (sector != 0) {
```

Comment:
Considering the fact, that failing to flash the CPU means something seriously wrong with the CPU itself, it is probably not worth it saving the first 16KB elsewhere and try to flash it back intact in case the preboot cannot be flashed (i.e. restore the previous bootloader)?

Reply:
I don't think so. At that point it is equally safe to flash the preboot or the original part of the bootloader. However, having the original part would add additional complexity to the code.

We might do some more thorough analysis to see all the possible reasons this could fail and see if we can improve its robustness.

One more thing I had on my mind (and I think would be nice to add eventually) is one more check of the bootloader data consistency (ensuring the buffer we are just about to flash to the CPU holds the data we expect it to hold). Something we can discuss in person.